### ISSUE TITLE
Scriptable Plugin API (C FFI)

### ISSUE THREAD

**Issue by NoraCodes** [xi-editor/xi-editor#1168]
Per [this proposal](https://xi-editor.io/xi-editor/gsoc.html#scriptable-language-bindings-c-api), I'd like to look at designing and implementing a scriptable, embeddable Xi API, through C bindings.

### The Problem

Performance is impacted by JSON-RPC round-tripping and could be greatly improved for most use cases by embedding the execution of xi-core in the editor GUI and communicating in-process via a FFI.

### The Solution

This solution will need to be implemented in 3 phases.

1. Designing a basic API and API framework, based on the critical functions present in the existing JSON-RPC API
2. Clearing any issues with an embeddable library version of xi-core
3. Implementing the embedded API alongside integration into one or more clients

**Comment by cmyr** [comment]
Hi Nora! there's currently some movement on this front already; @dsp has been experimenting with an embedded version of Xi that maintains the current API, and I am slowly hacking on a refactor that takes fuller advantage of being in process to offer certain API directly over FFI. If you're interested in this work I'm happy to talk more!

**Comment by NoraCodes** [comment]
Hi cmyr, I'm very interested in working on it. I'm currently writing up a proposal for GSoC regarding this.

**Comment by chriskrycho** [comment]
@cmyr if you get a chance to document how you all are tackling that, I'd find it super helpful as an interested observer—I'm working on an application for which Xi's architecture is very much an inspiration, and I imagine the same is true for others. If you don't have time, I of course totally understand, but even just a simple outline of the approach would be amazing!

**Comment by cmyr** [comment]
@chriskrycho This is very much ongoing work, but if there's anything in particular you're curious about I'm happy to answer Qs? Currently what I'm doing is just using direct function calls instead of RPC but am still passing JSON around in the general case, only using a more specific API in a few performance sensitive areas, such as width measurement (and some modal editing experiments).

**Comment by cmyr** [comment]
(my current tentative plan, which I haven't spoken much about, is to implement a mac-native 'rust playground' as a test-bed for some of these ideas; hope to make some real progress no that in the next couple of weeks.)

**Comment by chriskrycho** [comment]
Thanks – that's a helpful validation; I mostly just wanted to make sure it was that approach rather than using shared data structures more directly. I've been evaluating different approaches there in terms of overhead for UI and doing C FFI with probably-JSON payloads is exactly what I've landed on so far, so it's nice to know that *either* I'm not totally off my rocker *or* I at least have company. :wink:

If you do build that playground idea, I'd love to see it!

**Comment by cmyr** [comment]
So one thing i'm very curious about is the possibility of sharing data structures between rust libraries over FFI, and using them more directly. For general interop with other languages, though, this feels hard.

One thing you can do is pass C opaque pointers to rust structures, and then define FFI functions that take those pointers as arguments, convert them back to a rust type, and then call methods on them. So maybe this is closer to what you had in mind?

If you didn't see the [mozilla hacks post on FFI](https://hacks.mozilla.org/2019/04/crossing-the-rust-ffi-frontier-with-protocol-buffers/) I think it's quite good, and their [ffi-support crate](https://docs.rs/crate/ffi-support/0.3.2) has been a helpful reference.

**Comment by chriskrycho** [comment]
> For general interop with other languages, though, this feels hard.

Yeah, after looking at it a bunch, this was my conclusion as well. Doing Swift-Rust or Kotlin-Rust or .NET-Rust that way just seems like a recipe for pain.

> One thing you can do is pass C opaque pointers to rust structures, and then define FFI functions that take those pointers as arguments, convert them back to a rust type, and then call methods on them. So maybe this is closer to what you had in mind?

This was one of the options I'd considered, yeah, but…

> If you didn't see the mozilla hacks post on FFI I think it's quite good, and their ffi-support crate has been a helpful reference.

I did *not* see those; they're excellent, and have a bunch of options that are *exactly* what I was looking for. Thank you so much! (And now @NoraCodes I will stop derailing this thread!)

**Comment by NoraCodes** [comment]
@chriskrycho your derailment has been both useful and interesting, and provides useful insight on the use case for this feature!
